### PR TITLE
Automated cherry pick of #93044: Initialize scheduler's podInformer in sharedInformerFactory

### LINKED ISSUE #104521: Pods stuck in Pending status - kube-scheduler 1.19.13
#### What happened:
After upgrading core k8s components from 1.18 to 1.19.13 we faced a problem with Pending pods. From time to time PODs(different deployments) are stuck in Pending status. From kube-scheduler logs, we see a new event that starts appearing after the upgrade. For example `Pod kube-system/prometheus-thanos-frontend-86f4f8df77-h52tl **doesn't exist in informer cache**: pod "prometheus-thanos-frontend-86f4f8df77-h52tl" not found`.
However, ClusterAutoscaler still tries checking if there's an appropriate node `Pod kube-system.prometheus-thanos-frontend-86f4f8df77-h52tl marked as unschedulable can be scheduled on node ***.eu-west-1.compute.internal (based on hinting). Ignoring in scale up.`

It looks like kube-scheduler just forgot about the pod, because after a couple of attempts it stop trying to schedule the pod and we see no logs/events for hours. On the other hand, CA still tries to schedule it even spin up a new nodes.

#### What you expected to happen:

Pending pods are assigned to nodes.

#### How to reproduce it (as minimally and precisely as possible):

We don't see any pattern. It happens time to time for a completely different deployments.

#### Anything else we need to know?:

If we restart kube-scheduler so the leader is changed the Pending Pods are being successfully scheduled.

#### Environment:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.0", GitCommit:"af46c47ce925f4c4ad5cc8d1fca46c7b77d13b38", GitTreeState:"clean", BuildDate:"2020-12-08T17:59:43Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.13", GitCommit:"53c7b65d4531a749cd3a7004c5212d23daa044a9", GitTreeState:"clean", BuildDate:"2021-07-15T20:53:19Z", GoVersion:"go1.15.14", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration: AWS
- OS (e.g: `cat /etc/os-release`): CentOS Linux 7 (Core)
- Kernel (e.g. `uname -a`): Linux ***.eu-west-1.compute.internal 3.10.0-1160.25.1.el7.x86_64 #1 SMP Wed Apr 28 21:49:45 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
- Install tools:
- Network plugin and version (if this is a network-related bug): 
- Others: ClusterAutoscaler 1.19.1, kube-scheduler 1.19.13

### PR DESCRIPTION
**Pull request by alculquicondor**
Cherry pick of #93044 and #93566 on release-1.19.

#93044: Initialize scheduler's podInformer in sharedInformerFactory

#93566: Optimize the use of informer for scheduler

/sig scheduling
/kind bug

Because of the use of 2 separate informers, the scheduler might try to schedule a Pod, fail and never retry, because it thinks that the Pod doesn't exist. Reality was that the Pod was not yet added to the 2nd informer.

This is solved by using a single informer. This fix was originally introduced in 1.20 as a memory optimization, but it turns out to guarantee correct behavior.

#93566 is supposed to just be a refactoring. However, since we don't this would be the last 1.19 release, it's better to include it, as we have signal that both changes work well in production.

Fixes #104521

```release-note
Fix a regression in 1.19 where a race condition in scheduler could make a new unschedulable Pod never be retried.
```

### CODE REVIEW

**Comment by alculquicondor** [comment]
/assign @liggitt 
for test

/priority critical-urgent

File: pkg/scheduler/scheduler.go [hunk 1]
```diff
@@ -311,9 +306,6 @@ func initPolicyFromConfigMap(client clientset.Interface, policyRef *schedulerapi
 
 // Run begins watching and scheduling. It waits for cache to be synced, then starts scheduling and blocked until the context is done.
 func (sched *Scheduler) Run(ctx context.Context) {
-	if !cache.WaitForCacheSync(ctx.Done(), sched.scheduledPodsHasSynced) {
```
Comment (liggitt) [inline pkg/scheduler/scheduler.go:314]: we block for shared informers to sync before this point, right?

Comment (alculquicondor) [reply pkg/scheduler/scheduler.go:314]: yes, that's in cmd/kube-scheduler/app/server.go#Run

**Comment by liggitt** [comment]
does this code now match the state of 1.20, or were there follow up changes/fixes? ideally, this would take advantage of the signal we have in 1.20 (especially since this is the last 1.19 release), and https://github.com/kubernetes/kubernetes/pull/93044#issuecomment-678080923 sounds like https://github.com/kubernetes/kubernetes/pull/93566 also touched what shipped in 1.20 for this

**Comment by Huang-Wei** [comment]
@liggitt #93566 is just a cleanup, so doesn't quite need to be back ported.

**Comment by liggitt** [comment]
> @liggitt #93566 is just a cleanup, so doesn't quite need to be back ported.

we also thought https://github.com/kubernetes/kubernetes/pull/93044 was just an optimization... are we positive the change made in https://github.com/kubernetes/kubernetes/pull/93566 doesn't modify behavior in any way?

**Comment by justaugustus** [comment]
FYI @kubernetes/release-engineering 
ref: https://kubernetes.slack.com/archives/C2C40FMNF/p1631640903180200

/hold until discussion is resolved

File: pkg/scheduler/eventhandlers.go [hunk 2]
```diff
@@ -390,7 +388,7 @@ func addAllEventHandlers(
 		},
 	)
 	// unscheduled pod queue
-	podInformer.Informer().AddEventHandler(
+	informerFactory.Core().V1().Pods().Informer().AddEventHandler(
```
[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/scheduler/eventhandlers.go:391]: No, #96071 is not related to the use of the unified informer. The change was applied to deflake an integration test before we removed the syncPeriod in the test.

**Comment by alculquicondor** [comment]
I agree with @Huang-Wei  that #93566 was just supposed to be a refactoring. But we are safer to cherry-picking the change too as the intermediate state hasn't made it to any release and thus have no production signal.

**Comment by Huang-Wei** [comment]
> are we positive the change made in #93566 doesn't modify behavior in any way?

I read that #93566's key idea is to use a more well-encapsulated `NewFilteredPodInformer()` function to replace `cache.NewListWatchFromClient()`. Underneath they work the same. So I feel positive.

But I'm not going to oppose cherry-picking #93566 if we're more comfortable with containing it - it won't have side effects.

File: pkg/scheduler/eventhandlers.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (Huang-Wei) [reply pkg/scheduler/eventhandlers.go:391]: #96071 more deflakes an integration test that used a non-zero syncPeriod for infomers. In that case, we received old/new objects in update() with the same resourceVersion, and thus ends up with a flaking test.

In terms of regular logic, it's not impacted as it always uses a zero syncPeriod. Also, the latest integration test gets updated to use zero syncPeriod. So #96071 nows works as an extra gate checking.

[the remark this answers is not in the document]
Comment (Huang-Wei) [reply pkg/scheduler/eventhandlers.go:391]: Given it's not conventional to backport fixes of flaking test, I don't think we should backport #96701.

**Comment by alculquicondor** [comment]
I believe this should be merged in this patch cycle (1.19.15).
Without this patch, it is possible that some Pods remain unschedulable indefinitely. The only workaround possible is to restart the scheduler, which is hidden for most cloud providers.

**Comment by liggitt** [comment]
I agree. This is a regression from 1.18 that was accidentally resolved by this change that merged to 1.20. The regression needs to be resolved and this change has had soak time in 1.20 since 1.20.0

**Comment by justaugustus** [comment]
Approving this post-deadline cherry pick based on the comments from @alculquicondor and @liggitt:

> I believe this should be merged in this patch cycle (1.19.15).
> Without this patch, it is possible that some Pods remain unschedulable indefinitely. The only workaround possible is to restart the scheduler, which is hidden for most cloud providers.

> I agree. This is a regression from 1.18 that was accidentally resolved by this change that merged to 1.20. The regression needs to be resolved and this change has had soak time in 1.20 since 1.20.0

**Pull request merged**